File: packages/gavio-py/gavio/interceptors/cache/backends/memory.py
"""In-memory cache backend (F-CACHE-03) — default zero-dependency dev backend."""

from __future__ import annotations

import time
from collections import OrderedDict
from typing import Any

from ..backend import CacheBackend


class MemoryBackend(CacheBackend):
    """LRU-bounded, optionally TTL'd in-process cache. Not shared across processes."""

    def __init__(self, max_size: int = 1000) -> None:
        self.max_size = max_size
        self._store: OrderedDict[str, tuple[Any, float | None]] = OrderedDict()

    async def get(self, key: str) -> Any | None:
        entry = self._store.get(key)
        if entry is None:
            return None
        value, expires_at = entry
        if expires_at is not None and time.monotonic() > expires_at:
            del self._store[key]
            return None
        self._store.move_to_end(key)
        return value

    async def set(self, key: str, value: Any, ttl_seconds: int | None = None) -> None:
        expires_at = time.monotonic() + ttl_seconds if ttl_seconds else None
        self._store[key] = (value, expires_at)
        self._store.move_to_end(key)
        while len(self._store) > self.max_size:
            self._store.popitem(last=False)

    async def delete(self, key: str) -> None:
        self._store.pop(key, None)

    async def clear(self) -> None:
        self._store.clear()

    def __len__(self) -> int:
        return len(self._store)

File: packages/gavio-js/src/interceptors/cache/backends/memory.ts
/** In-memory cache backend (F-CACHE-03) — default zero-dependency dev backend. */

import type { CacheBackend } from '../backend.js'

interface Entry {
  value: unknown
  expiresAt: number | null
}

export interface MemoryCacheBackendOptions {
  maxSize?: number
}

/** LRU-bounded, optionally TTL'd in-process cache. Not shared across processes. */
class MemoryBackend implements CacheBackend {
  readonly maxSize: number
  // Map preserves insertion order, which we use for LRU eviction.
  private store = new Map<string, Entry>()

  constructor(maxSize = 1000) {
    this.maxSize = maxSize
  }

  async get(key: string): Promise<unknown | null> {
    const entry = this.store.get(key)
    if (entry === undefined) return null
    if (entry.expiresAt !== null && now() > entry.expiresAt) {
      this.store.delete(key)
      return null
    }
    // Move to end (most-recently-used).
    this.store.delete(key)
    this.store.set(key, entry)
    return entry.value
  }

  async set(key: string, value: unknown, ttlSeconds?: number | null): Promise<void> {
    const expiresAt = ttlSeconds ? now() + ttlSeconds * 1000 : null
    this.store.delete(key)
    this.store.set(key, { value, expiresAt })
    while (this.store.size > this.maxSize) {
      const oldest = this.store.keys().next().value
      if (oldest === undefined) break
      this.store.delete(oldest)
    }
  }

  async delete(key: string): Promise<void> {
    this.store.delete(key)
  }

  async clear(): Promise<void> {
    this.store.clear()
  }

  get size(): number {
    return this.store.size
  }
}

function now(): number {
  return Date.now()
}

/** Factory: build an in-memory cache backend. */
export function memoryCacheBackend(options: MemoryCacheBackendOptions = {}): CacheBackend {
  return new MemoryBackend(options.maxSize ?? 1000)
}
